tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php

```diff
@@ -863,6 +864,84 @@ private function skipOci8AndMysqli()
             $this->markTestSkipped('Mysqli driver dont support this feature.');
         }
     }
+
+    public function testIteratorCallsFetchTheExpectedNumberOfTimes()
+    {
+        $sql = 'SELECT CURTIME(6) AS time_started, test_int, test_string FROM fetch_table';
+        $originalStmt = $this->_conn->prepare($sql);
+
+        // Mock the connection
+        $connectionMock = $this->getMockBuilder('Doctrine\\DBAL\\Driver\\Connection')
+            ->setMethods(['prepare', 'query'])
+            ->getMockForAbstractClass();
+
+        $stmtSpy = new StatementSpy($originalStmt);
```

Comment:
> But we can test all of them in the same suite using data provider

I'm not sure about this, as not every driver will get loaded you could end up with errors regarding missing class constants etc.

> I don't know how to invoke getIterator() implicitly without calling current()/next() and without depending on the actual implementation

This is why I created a spy manually. I tried the inner iterator thing you mentioned but couldn't get it to work.

Reply:
> I'm not sure about this, as not every driver will get loaded you could end up with errors regarding missing class constants etc.

On the mocked statement classes, the test will only call the `getIterator()` methods which do not use any driver-specific constants.

> This is why I created a spy manually. I tried the inner iterator thing you mentioned but couldn't get it to work.

But you still can call  `getIterator()` explicitly as in the example above. I think it's not a big problem.

Reply:
The following test throws up the error `1) Doctrine\Tests\DBAL\Functional\DataAccessTest::testGettingIteratorDoesNotCallFetch with data set #1 ('Doctrine\DBAL\Driver\OCI8\OCI...ection')
Use of undefined constant OCI_COMMIT_ON_SUCCESS - assumed 'OCI_COMMIT_ON_SUCCESS'`

Admittedly this is likely because I don't have the OCI8 extension loaded, but then seems a shame to have to have all extensions loaded to run unit tests against just one driver.

Running the test with `vendor/bin/phpunit --filter DataAccessTest::testGettingIteratorDoesNotCallFetch --configuration tests/travis/mysqli.travis.xml`

```php
    public function getConnections()
    {
        return [
            [MysqliConnection::class],
            [OCI8Connection::class],
        ];
    }

    /**
     * @dataProvider getConnections
     */
    public function testGettingIteratorDoesNotCallFetch($connection)
    {
        $stmt = $this->createMock(Statement::class);
        $stmt->expects($this->never())->method('fetch');

        $conn = $this->createMock($connection);
        $conn->expects($this->any())->method('prepare')->willReturn($stmt);

        $ii = new IteratorIterator($stmt);
        $ii->getInnerIterator();
    }
```

> But you still can call getIterator() explicitly as in the example above. I think it's not a big problem.

Will need some extra clarification about this one. You can call `getIterator()` but it's on a mock, so doesn't return anything? Only way I can see of doing it is by basically re-writing the implementation within the mock?

Reply:
> The following test throws up the error `1) Doctrine\Tests\DBAL\Functional\DataAccessTest::testGettingIteratorDoesNotCallFetch with data set #1 ('Doctrine\DBAL\Driver\OCI8\OCI...ection')
Use of undefined constant OCI_COMMIT_ON_SUCCESS - assumed 'OCI_COMMIT_ON_SUCCESS'`

Could you try moving this to constructor?
 https://github.com/doctrine/dbal/blob/4da603c97002c4a09f271d96bf808e62d8a3b9dc/lib/Doctrine/DBAL/Driver/OCI8/OCI8Connection.php#L41

> Will need some extra clarification about this one. You can call `getIterator()` but it's on a mock, so doesn't return anything? Only way I can see of doing it is by basically re-writing the implementation within the mock?

`getIterator()` shouldn't be mocked since we're testing it. Please try using something else than `createMock()` — this one will mock all methods. `createPartialMock()` may be a good alternative.

Please feel free to commit the failing test so that I could offer a fix if you're tired of rewriting it.

Reply:
How did I not know about partial mocks?! 😩 

Ok this helps a lot, I'm still struggling a bit with the second test. Have pushed it up.